### PR TITLE
implement dashboard service

### PR DESCRIPTION
**Pull request by stat-kwon**
### Category
- [x] New feature
- [ ] Bug fix
- [ ] Improvement
- [ ] Refactor
- [ ] etc

### Description
A new dashboard proto buffer was created according to the dashboard's api document. The dashboard service includes the following resources.
- DomainDashboard
- ProjectDashboard
- Widget

### CODE REVIEW

**Review by whdalsrnt — CHANGES_REQUESTED** [review]
You need to change the protobuf spec.

File: proto/spaceone/api/dashboard/v1/domain_dashboard.proto [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+  // is_required: false
+  spaceone.api.core.v1.Query query = 1;
+  // is_required: false
+  string domain_dashboard_id = 2;
+  // is_required: false
+  string name = 3;
+  // is_required: false
+  Scope scope = 4;
+  // is_required: false
+  google.protobuf.ListValue labels = 5;
```
Comment (whdalsrnt) [inline proto/spaceone/api/dashboard/v1/domain_dashboard.proto:137]: Remove the `labels` filter

File: proto/spaceone/api/dashboard/v1/project_dashboard.proto [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+  // is_required: false
+  spaceone.api.core.v1.Query query = 1;
+  // is_required: false
+  string project_dashboard_id = 2;
+  // is_required: false
+  string name = 3;
+  // is_required: false
+  Scope scope = 4;
+  // is_required: false
+  google.protobuf.ListValue labels = 5;
```
Comment (whdalsrnt) [inline proto/spaceone/api/dashboard/v1/project_dashboard.proto:137]: Remove the `labels` filter

File: proto/spaceone/api/dashboard/v1/domain_dashboard.proto [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  string start = 1;
+  string end = 2;
+}
+
+message DomainDashboardCurrency {
+  bool enabled = 1;
+}
+
+message CreateDomainDashboardRequest {
+  enum Scope {
```
Comment (whdalsrnt) [inline proto/spaceone/api/dashboard/v1/domain_dashboard.proto:66]: Add NONE type as follows:
- SCOPE_NONE
- DOMAIN
- USER

File: proto/spaceone/api/dashboard/v1/domain_dashboard.proto [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  // is_required: true
+  string domain_dashboard_id = 1;
+  // is_required: true
+  string domain_id = 2;
+  // is_required: false
+  repeated string only = 3;
+}
+
+message DomainDashboardQuery {
+  enum Scope {
```
Comment (whdalsrnt) [inline proto/spaceone/api/dashboard/v1/domain_dashboard.proto:124]: Add NONE type as follows:
- SCOPE_NONE
- DOMAIN
- USER

File: proto/spaceone/api/dashboard/v1/domain_dashboard.proto [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  // is_required: false
+  google.protobuf.ListValue labels = 5;
+  // is_required: false
+  string user_id = 10;
+  // is_required: true
+  string domain_id = 11;
+}
+
+message DomainDashboardInfo {
+  enum Scope {
```
Comment (whdalsrnt) [inline proto/spaceone/api/dashboard/v1/domain_dashboard.proto:145]: Add NONE type as follows:
- SCOPE_NONE
- DOMAIN
- USER

File: proto/spaceone/api/dashboard/v1/project_dashboard.proto [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+  string start = 1;
+  string end = 2;
+}
+
+message ProjectDashboardCurrency {
+  bool enabled = 1;
+}
+
+message CreateProjectDashboardRequest {
+  enum Scope {
```
Comment (whdalsrnt) [inline proto/spaceone/api/dashboard/v1/project_dashboard.proto:66]: Add NONE type as follows:
- SCOPE_NONE
- DOMAIN
- USER

File: proto/spaceone/api/dashboard/v1/project_dashboard.proto [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+  // is_required: true
+  string project_dashboard_id = 1;
+  // is_required: true
+  string domain_id = 2;
+  // is_required: false
+  repeated string only = 3;
+}
+
+message ProjectDashboardQuery {
+  enum Scope {
```
Comment (whdalsrnt) [inline proto/spaceone/api/dashboard/v1/project_dashboard.proto:124]: Add NONE type as follows:
- SCOPE_NONE
- DOMAIN
- USER

File: proto/spaceone/api/dashboard/v1/project_dashboard.proto [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+  // is_required: false
+  google.protobuf.ListValue labels = 5;
+  // is_required: false
+  string user_id = 10;
+  // is_required: true
+  string domain_id = 11;
+}
+
+message ProjectDashboardInfo {
+  enum Scope {
```
Comment (whdalsrnt) [inline proto/spaceone/api/dashboard/v1/project_dashboard.proto:145]: Add NONE type as follows:
- SCOPE_NONE
- DOMAIN
- USER

File: proto/spaceone/api/dashboard/v1/widget.proto [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+      }
+    };
+  }
+
+  rpc stat (WidgetStatQuery) returns (google.protobuf.Struct) {
+    option (google.api.http) = {post: "/dashboard/v1/widgets/stat"};
+  }
+}
+
+enum ViewMode {
```
Comment (whdalsrnt) [inline proto/spaceone/api/dashboard/v1/widget.proto:45]: Add NONE type as follows:
- VIEW_MODE_NONE
- AUTO
- FULL

**Review by whdalsrnt — CHANGES_REQUESTED** [review]

**Comment by stat-kwon** [comment]
There were 2 fixes based on code review.
- Removed labels field from query.
- Added None type to scope.

File: proto/spaceone/api/dashboard/v1/domain_dashboard.proto (hunk 3 above)
Comment (whdalsrnt) [reply proto/spaceone/api/dashboard/v1/domain_dashboard.proto:66]: The order of `NONE` values is set to 0.

File: proto/spaceone/api/dashboard/v1/domain_dashboard.proto (hunk 4 above)
Comment (whdalsrnt) [reply proto/spaceone/api/dashboard/v1/domain_dashboard.proto:125]: The order of `NONE` values is set to 0.

File: proto/spaceone/api/dashboard/v1/domain_dashboard.proto (hunk 5 above)
Comment (whdalsrnt) [reply proto/spaceone/api/dashboard/v1/domain_dashboard.proto:145]: The order of `NONE` values is set to 0.

File: proto/spaceone/api/dashboard/v1/project_dashboard.proto (hunk 6 above)
Comment (whdalsrnt) [reply proto/spaceone/api/dashboard/v1/project_dashboard.proto:66]: The order of `NONE` values is set to 0.

File: proto/spaceone/api/dashboard/v1/project_dashboard.proto (hunk 7 above)
Comment (whdalsrnt) [reply proto/spaceone/api/dashboard/v1/project_dashboard.proto:125]: The order of `NONE` values is set to 0.

File: proto/spaceone/api/dashboard/v1/project_dashboard.proto (hunk 8 above)
Comment (whdalsrnt) [reply proto/spaceone/api/dashboard/v1/project_dashboard.proto:145]: The order of `NONE` values is set to 0.

File: proto/spaceone/api/dashboard/v1/widget.proto (hunk 9 above)
Comment (whdalsrnt) [reply proto/spaceone/api/dashboard/v1/widget.proto:45]: The order of `NONE` values is set to 0.

**Review by whdalsrnt — CHANGES_REQUESTED** [review]

**Comment by WANZARGEN** [comment]
Looks good to me !!!!

**Comment by stat-kwon** [comment]
Overall protobuf change according to API document change in dashboard service
- Changed terminology and redesigned the overall structure
- Add project_id to project dashboard

**Comment by stat-kwon** [comment]
We changed the terminology because widget_id is ambiguous from the point of view of api users.

AS-IS
- `widget_id`
- `name`

TO-BE
- `widget_name`
- `title`

Code review please. @whdalsrnt @jihyungSong

**Comment by stat-kwon** [comment]
The following API has been added to add the version control function to the dashboard service.

**[DomainDashboard]**

- verb
        - DomainDashboard.delete_version
        - DomainDashboard.revert_version
        - DomainDashboard.get_version 
        - DomainDashboard.list_versions

- info
    -DomainDashboardVersionInfo


**[ProjectDashboard]**

- verb
        - ProjectDashboard.delete_version
        - ProjectDashboard.revert_version
        - ProjectDashboard.get_version 
        - ProjectDashboard.list_versions

- info
        - ProjectDashboardVersionInfo

**Review by whdalsrnt — APPROVED** [review]
It looks perfect code to me.

**Pull request merged**